src/main/scala/is/hail/utils/Graph.scala

```diff
@@ -0,0 +1,53 @@
+package is.hail.utils
+
+import scala.collection.mutable
+import scala.reflect.ClassTag
+
+object Graph {
+  def mkGraph[T](edges: (T, T)*): mutable.MultiMap[T, T] =
+    mkGraph(edges.toArray)
+
+  def mkGraph[T](edges: Array[(T, T)]): mutable.MultiMap[T, T] = {
+    val m = new mutable.HashMap[T, mutable.Set[T]]() with mutable.MultiMap[T, T]
+    var i = 0
+    while (i < edges.length) {
+      m.addBinding(edges(i)._1, edges(i)._2)
+      m.addBinding(edges(i)._2, edges(i)._1)
+      i += 1
+    }
+    m
+  }
+
+  def mkGraph[T](edges: TraversableOnce[(T, T)]): mutable.MultiMap[T, T] = {
+    val m = new mutable.HashMap[T, mutable.Set[T]]() with mutable.MultiMap[T, T]
+    edges.foreach { case (i, j) =>
+      m.addBinding(i, j)
+      m.addBinding(j, i)
+    }
+    m
+  }
+
+  def maximalIndependentSet[T: ClassTag](edges: Array[(T, T)]): Array[T] = {
+    maximalIndependentSet(mkGraph(edges))
+  }
+
+  def maximalIndependentSet[T: ClassTag](g: mutable.MultiMap[T, T]): Array[T] = {
+    val verticesByDegree = new BinaryHeap[T]()
+
+    g.foreach { case (v, neighbors) =>
+      verticesByDegree.insert(v, neighbors.size)
+    }
+
+    while (verticesByDegree.nonEmpty && verticesByDegree.maxPriority() > 0) {
+      val current = verticesByDegree.extractMax()
+      val neighbors = g(current)
+      neighbors.foreach { x =>
+        if (verticesByDegree.contains(x))
```

Comment:
A thought on implementation: when you extract x, you could also remove x from the neighbor set of it's neighbors. I expect this would be slower when removing a small fraction of vertices, but possibly faster when removing the majority of vertices in a highly connected graph since the foreach is over gradually smaller sets and `verticesByDegree.contains(x)` need not be checked. Since we expect low edge density in our current applications, I'll leave it to you whether this is worth investigating (also given your understanding of where time is being spent overall)

Reply:
I ran some mild tests and it seems that removing from the neighbor set is as fast and maybe faster for 1M mother-father-child triplets. I'll make the change.